### PR TITLE
Building an expenses app with Ruby on Rails (All In One )

### PR DESCRIPTION
**Pull request by youmari**
### ⚠️ I didn't implement the required font cause it's not for free you have to purchase it 
- Splash screen
  - A simple page with the name of the app, and links to the sign up and log in pages.

- Sign up and log in pages
  - The user should be able to register in the app with their full name, email, and password (all mandatory).
  - The user can log into the app using email and password.
  - If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

- Home page (catagories page)
  - When the user logs in, they are presented with the categories page.
  - For each category, the user can see their name, icon, and the total amount of all the transactions that belongs to that category.
  - When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
  - There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

- Expenses page
  - For a given category, the list of Expenses is presented, ordered by the most recent.
  - At the top of the page the user could see the total amount for the category (sum of all of the amounts of the Expenses in that category).
  - There is a button "add a new Expense" at the bottom that brings the user to the page to create a new Expense.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new category" page
  - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
  - The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new Expense" page
  - The user fills out a form to create a new Expense with:
    - name (mandatory)
    - amount (mandatory)
    - categories (mandatory at least one)
  - The user clicks (or taps) the "Save" button to create the new category, and is taken to the Expenses page for that category.
  - When the user clicks on the "Back" button (<), the user navigates to the Expenses page for that category.

#### Testing requirements
- Create unit and integration tests for all the most important components of the All in One application.

### CODE REVIEW

**Comment by youmari** [comment]
## 🎥 [Loom video](https://www.loom.com/share/dd30fbbeca5a4790848726e4bd5359fd) of me describing the app
https://www.loom.com/share/dd30fbbeca5a4790848726e4bd5359fd

File: README.md [hunk 1]
```diff
@@ -1 +1,105 @@
...
+
+## Acknowledgments
+
+- Special thanks for [Gregoire Vella on Behance](https://www.behance.net/gregoirevella) the Owner of the Original design
+
+## Show your support
+
+Give a ⭐️ if you like this project!
+
+Thanks
```
Comment (FrederickMih) [inline README.md]: - [ ] Your README file is descriptive :+1: . Please make it professional by including the License section. :smiley:

File: app/controllers/expenses_controller.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  # Use callbacks to share common setup or constraints between actions.
+  def set_expense
+    @expense = Expense.find(params[:id])
+  end
+
+  # Only allow a list of trusted parameters through.
+  def expense_params
+    params.require(:expense).permit(:name, :amount, expense: [:group_id])
+  end
+end
```
Comment (FrederickMih) [inline app/controllers/expenses_controller.rb:70]: - [ ] Hi @youmari, I think you will require corresponding views for the update and destroy actions. Please let's consider this.

File: app/controllers/groups_controller.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  # Use callbacks to share common setup or constraints between actions.
+  def set_group
+    @group = Group.find(params[:id])
+  end
+
+  # Only allow a list of trusted parameters through.
+  def group_params
+    params.require(:group).permit(:name, :icon)
+  end
+end
```
Comment (FrederickMih) [inline app/controllers/groups_controller.rb:70]: - [ ] I can't update or delete a group. Please make sure that these features are functional in your view. These actions have been added here but no view is available for that.

**Review by FrederickMih — CHANGES_REQUESTED** [review]
Hi @youmari  , 

You have done an incredible job so far. Keep up! :rocket: 
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!

Screenshot from 2022-04-12 20-44-47

# To Highlights

-  Git/GitHub workflow has been followed :heavy_check_mark: 
-  Linters are correctly set up :heavy_check_mark: 
-  Pull Request has a title and summary :heavy_check_mark: 

## Suggested changes

- [ ] **Video presentation**: Hi @youmari  , the video presentation of your project is missing some requirements:
- The video should contain a highlight about a piece of the code you find interesting or proud of making.
It will be nice if you can record another video to include the requirements above. You ended up presenting the features of the project on a browser. You didn't open up the IDE.

- [ ] Hi @youmari, I think you will require to display the created date of an expense and a group. I will be nice to keep track of the date on which an expense was created.
- [ ] Most of your tests are failing. Please make sure to fix them.


_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: app/controllers/expenses_controller.rb (hunk 2 above)
Comment (youmari) [reply app/controllers/expenses_controller.rb:70]: @FrederickMih the views for update and destroy are not required

File: app/controllers/groups_controller.rb (hunk 3 above)
Comment (youmari) [reply app/controllers/groups_controller.rb:70]: @FrederickMih but these are not in the requirements of this project

Comment (youmari) [reply app/controllers/groups_controller.rb:70]: @FrederickMih also for the video they didn't say you have to open the ide 
they say: 
- description of the project 
    - which I did I described the project  in the video
 - a demo of the project 
     - which is also I did it in the video

Comment (FrederickMih) [reply app/controllers/groups_controller.rb:70]: @youmari, how are you going to highlight the piece of code you found interesting or proud of?. I think you need to open your IDE. I can see that you are only interested in pointing out what you think you did correctly while ignoring your faults.

File: app/controllers/expenses_controller.rb (hunk 2 above)
Comment (FrederickMih) [reply app/controllers/expenses_controller.rb:70]: @youmari, they are not required then you shouldn't have added the actions :angry:

**Comment by youmari** [comment]
@FrederickMih I talked with a senior code reviewer about the changes that you requested me that's what he said : 
image

**Comment by youmari** [comment]
## Another loom video 
https://www.loom.com/share/de44bbf3c8174201a3960fbdab172924

**Comment by ShoiraTa** [comment]
## **STATUS: APPROVED  🟢 ** 
------
### Hi @youmari , 

Your project is complete! . There is nothing else to say other than... it's time to merge it 🍾🚢 :shipit: . Congratulations!  🎉💯🌟 

5SW

**Cheers and Happy coding! :clap: :clap: :clap:**

Feel free to leave any questions or comments in the PR thread by mentioning my Github username if something is not 100% clear.
***
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have 2 more limited reviews per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**